tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js

```diff
@@ -168,13 +164,71 @@ TcpSync.prototype._readResponse = function() {
     throw error;
   }
 
-  return data;
+  var res;
+
+  if (this.marionetteProtocol >= 3) {
+    var resp = Response.fromMsg(data);
+    res = resp.error || resp.result;
+  } else {
+    res = data;
+  }
+
+  return res;
```

Comment:
If all you're doing is returning `res` you could just return above and avoid nesting